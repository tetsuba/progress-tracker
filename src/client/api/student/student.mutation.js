import gql from 'graphql-tag'

export const ADD_STUDENT_MUTATION = gql`
  mutation AddStudent($input: AddStudentInput!) {
    addStudent(input: $input) {
      _id
      firstName
      lastName
    }
  }
`
