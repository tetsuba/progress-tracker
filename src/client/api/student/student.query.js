import gql from 'graphql-tag'

export const STUDENTS_QUERY = gql`
  {
    students {
      id
      firstName
      lastName
    }
  }
`

export const STUDENT_QUERY = gql`
  query($input: StudentIdInput!) {
    getStudent(input: $input) {
      firstName
      lastName
      courses
    }
  }
`
