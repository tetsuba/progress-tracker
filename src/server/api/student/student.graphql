
type StudentType {
  id: String
  firstName: String
  lastName: String
  DOB: String
  courses: [String]
}

type StudentsType {
  id: String
  firstName: String
  lastName: String
}

input StudentIdInput {
  id: String
}

input StudentCourseInput {
  studentId: String
  courseName: String,
}

input AddStudentInput {
  firstName: String
  lastName: String
  DOB: String
}

extend type Query {
  students: [StudentsType]!
  getStudent(input: StudentIdInput!): StudentType!
}

extend type Mutation {
  addStudent(input: AddStudentInput!): [StudentsType]!
  addStudentCourse(input: StudentCourseInput!): StudentType!
}
