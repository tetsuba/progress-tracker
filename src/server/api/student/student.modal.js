const mongoose = require('mongoose')

// TODO: Create user schema -- No id. Is this correct?

const studentSchema = new mongoose.Schema({
  firstName: {
    type: String,
    required: true,
  },
  lastName: {
    type: String,
    required: true,
  },
  DOB: {
    type: String,
    required: true,
  },
  teacherID: {
    type: String,
    required: true,
  },
})

mongoose.model('student', studentSchema)
