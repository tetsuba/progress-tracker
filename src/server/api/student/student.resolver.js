const { errorName } = require('../../errorHandling')

module.exports = {
  Query: {
    students: async (_, args, context) => {
      const { id } = context.user
      const { Student } = context.models
      return await Student.find().where('teacherID').in(id).exec()
    },
    getStudent: (_, args, context) => {
      const { Student } = context.models
      const { id } = args.input
      return Student.findOne({ _id: id })
    },
  },

  Mutation: {
    addStudent: async (_, args, context, info) => {
      const { id } = context.user
      const { Student } = context.models
      const student = await Student.findOne({ firstName: args.input.firstName })

      // TODO: This is not practical and require to rethink what to do here
      if (student) {
        throw new Error(errorName.STUDENT_EXISTS)
      }

      await Student.create({
        ...args.input,
        teacherID: id,
      })

      return await Student.find().where('teacherID').in(id).exec()
    },
    addStudentCourse: async (_, args, context, info) => {
      const { Student, CourseABC } = context.models
      const { studentId, courseName } = args.input

      const student = await Student.findById(studentId)
      student.courses.push(courseName)
      await student.save()

      // TODO: check for student id does not exist before creating a new Course ABC
      await CourseABC.create({
        studentId: studentId,
      })

      return student
    },
  },
}
