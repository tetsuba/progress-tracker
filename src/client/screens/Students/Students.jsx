import React, { useContext, useEffect } from 'react'
import { Link } from 'react-router-dom'
import { useQuery } from '@apollo/react-hooks'
import { Button, Col, Container, ListGroup, Row } from 'react-bootstrap'

// COMPONENTS
// $FlowFixMe - Investigate how to fix flow errors
import AddStudentModal from '../../components/Modal/AddStudentModal'
import BreadCrumbs from '../../components/BreadCrumbs/BreadCrumbs'
import { CRUMBS_KEY } from '../../components/BreadCrumbs/crumbs'

// ICONS
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { faPlus } from '@fortawesome/free-solid-svg-icons'

// CONTEXT
// $FlowFixMe - Investigate how to fix flow errors
import { ModalContext } from '../../context/ModalContext'

// QUERY
import { STUDENTS_QUERY } from '../../api/student/student.query'

export default function Students() {
  // $FlowFixMe - Graphql investigate how to fix
  const { loading, data } = useQuery(STUDENTS_QUERY)
  const { toggleModal, addTemplate } = useContext(ModalContext)

  useEffect(() => {
    addTemplate(AddStudentModal)
  }, [addTemplate])

  const renderList = (students) =>
    students.map((student) => (
      <Link
        to={`/student/${student._id}`}
        key={student._id}
        className="list-group-item"
      >
        {student.firstName} {student.lastName}
      </Link>
    ))

  return (
    <Container>
      <Row className="mt-5">
        <BreadCrumbs crumbKey={CRUMBS_KEY.STUDENTS} />
      </Row>
      <Row className="mt-5">
        <Col>
          <h1>My Students</h1>
        </Col>
      </Row>
      <Row className="mt-5">
        <Col>
          <Button
            className="float-right"
            variant="primary"
            onClick={() => toggleModal()}
          >
            <FontAwesomeIcon icon={faPlus} /> Add Student
          </Button>
        </Col>
      </Row>
      <Row className="mt-1">
        <Col>
          <ListGroup>{!loading && renderList(data.students)}</ListGroup>
        </Col>
      </Row>
    </Container>
  )
}
