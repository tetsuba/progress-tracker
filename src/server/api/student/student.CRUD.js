const mongoose = require('mongoose');
require('./student.modal');
const Student = mongoose.model('student');

function findAllStudents() {
    return Student.find({})
}

async function getStudentById(data) {
    const std = await Student.findOne({_id: data.id});
    return std;
}

async function createNewStudent(data) {
    const student = await Student.findOne({ firstName: data.firstName});

    if(student) {
        // TODO: how to handle this error better.
        return new Error('Student exists')
    }
    const newStudent = await Student.create(data);
    return { ...newStudent, id: newStudent._id };
}


module.exports = {
    findAllStudents,
    createNewStudent,
    getStudentById,
};
