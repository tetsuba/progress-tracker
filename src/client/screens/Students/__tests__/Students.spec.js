import { act } from 'react-dom/test-utils'
import { delay, graphRenderer } from '../../../../test/testHelper'
import Students from '../Students'
import { getStudentsQuerySuccess } from '../../../../test/mockApi/student/studentMockQuery'

jest.mock('react-router-dom', () => ({
  Link: ({ children }) => children,
}))

describe('Students', () => {
  describe('@Render', () => {
    it('Should render', async () => {
      let wrapper
      await act(async () => {
        wrapper = graphRenderer(Students, [getStudentsQuerySuccess], {})
        await delay()
      })
      wrapper.update()
      expect(wrapper.find(Students)).toMatchSnapshot()
    })
  })
})
