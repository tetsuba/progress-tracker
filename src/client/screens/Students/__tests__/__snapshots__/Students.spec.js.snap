// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Students @Render Should render 1`] = `
<Students>
  <Container
    fluid={false}
  >
    <div
      className="container"
    >
      <Row
        className="mt-5"
        noGutters={false}
      >
        <div
          className="mt-5 row"
        >
          <BreadCrumbs
            crumbKey="Students"
          >
            <Row
              noGutters={false}
            >
              <div
                className="row"
              >
                <Breadcrumb
                  label="breadcrumb"
                  listProps={Object {}}
                >
                  <nav
                    aria-label="breadcrumb"
                  >
                    <ol
                      className="breadcrumb"
                    >
                      <Crumbs>
                        <li
                          className="breadcrumb-item"
                          key="Home-0"
                        >
                          <Link
                            to="/"
                          >
                            Home
                          </Link>
                        </li>
                        <BreadcrumbItem
                          active={true}
                          key="Students-1"
                          linkProps={Object {}}
                        >
                          <li
                            aria-current="page"
                            className="breadcrumb-item active"
                          >
                            Students
                          </li>
                        </BreadcrumbItem>
                      </Crumbs>
                    </ol>
                  </nav>
                </Breadcrumb>
              </div>
            </Row>
          </BreadCrumbs>
        </div>
      </Row>
      <Row
        className="mt-5"
        noGutters={false}
      >
        <div
          className="mt-5 row"
        >
          <Col>
            <div
              className="col"
            >
              <h1>
                My Students
              </h1>
            </div>
          </Col>
        </div>
      </Row>
      <Row
        className="mt-5"
        noGutters={false}
      >
        <div
          className="mt-5 row"
        >
          <Col>
            <div
              className="col"
            >
              <Button
                active={false}
                className="float-right"
                disabled={false}
                onClick={[Function]}
                type="button"
                variant="primary"
              >
                <button
                  className="float-right btn btn-primary"
                  disabled={false}
                  onClick={[Function]}
                  type="button"
                >
                  <FontAwesomeIcon
                    border={false}
                    className=""
                    fixedWidth={false}
                    flip={null}
                    icon={
                      Object {
                        "icon": Array [
                          448,
                          512,
                          Array [],
                          "f067",
                          "M416 208H272V64c0-17.67-14.33-32-32-32h-32c-17.67 0-32 14.33-32 32v144H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h144v144c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32V304h144c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z",
                        ],
                        "iconName": "plus",
                        "prefix": "fas",
                      }
                    }
                    inverse={false}
                    listItem={false}
                    mask={null}
                    pull={null}
                    pulse={false}
                    rotation={null}
                    size={null}
                    spin={false}
                    swapOpacity={false}
                    symbol={false}
                    title=""
                    transform={null}
                  >
                    <svg
                      aria-hidden="true"
                      className="svg-inline--fa fa-plus fa-w-14 "
                      data-icon="plus"
                      data-prefix="fas"
                      focusable="false"
                      role="img"
                      style={Object {}}
                      viewBox="0 0 448 512"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M416 208H272V64c0-17.67-14.33-32-32-32h-32c-17.67 0-32 14.33-32 32v144H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h144v144c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32V304h144c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
                        fill="currentColor"
                        style={Object {}}
                      />
                    </svg>
                  </FontAwesomeIcon>
                   Add Student
                </button>
              </Button>
            </div>
          </Col>
        </div>
      </Row>
      <Row
        className="mt-1"
        noGutters={false}
      >
        <div
          className="mt-1 row"
        >
          <Col>
            <div
              className="col"
            >
              <ListGroup
                horizontal={null}
                variant={null}
              >
                <ForwardRef
                  as="div"
                  className="list-group"
                  onSelect={[Function]}
                >
                  <div
                    className="list-group"
                    onKeyDown={[Function]}
                  >
                    <Link
                      className="list-group-item"
                      key="ID001"
                      to="/student/ID001"
                    >
                      FirstName001
                       
                      LastName001
                    </Link>
                    <Link
                      className="list-group-item"
                      key="ID002"
                      to="/student/ID002"
                    >
                      FirstName002
                       
                      LastName002
                    </Link>
                    <Link
                      className="list-group-item"
                      key="ID003"
                      to="/student/ID003"
                    >
                      FirstName003
                       
                      LastName003
                    </Link>
                  </div>
                </ForwardRef>
              </ListGroup>
            </div>
          </Col>
        </div>
      </Row>
    </div>
  </Container>
</Students>
`;
